pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go

```diff
@@ -235,32 +250,89 @@ func nodesWherePreemptionMightHelp(nodes []*framework.NodeInfo, m framework.Node
+// dry running preemption so that the same set of nodes don't get preempted
+// every time. If <offset> + <numCandidates> runs past the number of nodes, the
+// caller is expected to wrap around the node list.
+func (pl *DefaultPreemption) calculateOffset(numNodes int32) (int32, int32) {
+	offset := pl.candidateOffset
+	numCandidates := pl.calculateNumCandidates(numNodes)
+	if pl.candidateOffset >= numNodes {
+		// This happens when the cluster is downscaled and candidateOffset is past
+		// the new, reduced number of nodes. Reset the counter and restart from the
+		// beginning.
+		pl.candidateOffset = 0
+	} else {
+		pl.candidateOffset = (pl.candidateOffset + numCandidates) % numNodes
+	}
+	return offset, numCandidates
+}
+
 // dryRunPreemption simulates Preemption logic on <potentialNodes> in parallel,
-// and returns all possible preemption candidates.
-func dryRunPreemption(ctx context.Context, fh framework.PreemptHandle, state *framework.CycleState,
-	pod *v1.Pod, potentialNodes []*framework.NodeInfo, pdbs []*policy.PodDisruptionBudget) []Candidate {
-	var resultLock sync.Mutex
-	var candidates []Candidate
+// and returns preemption candidates. The number of candidates depends on the
+// constraints defined in the plugin's args. In the returned list of
+// candidates, ones that do not violate PDB are preferred over ones that do.
+func (pl *DefaultPreemption) dryRunPreemption(ctx context.Context, fh framework.PreemptHandle,
+	state *framework.CycleState, pod *v1.Pod, potentialNodes []*framework.NodeInfo,
+	pdbs []*policy.PodDisruptionBudget) []Candidate {
+	offset, numCandidates := pl.calculateOffset(int32(len(potentialNodes)))
+	var mu sync.Mutex
+	var nonViolatingCandidates []Candidate
+	var violatingCandidates []Candidate
+	parallelCtx, cancel := context.WithCancel(ctx)
 
 	checkNode := func(i int) {
-		nodeInfoCopy := potentialNodes[i].Clone()
+		nodeInfoCopy := potentialNodes[(int(offset)+i)%len(potentialNodes)].Clone()
 		stateCopy := state.Clone()
 		pods, numPDBViolations, fits := selectVictimsOnNode(ctx, fh, stateCopy, pod, nodeInfoCopy, pdbs)
 		if fits {
-			resultLock.Lock()
 			victims := extenderv1.Victims{
 				Pods:             pods,
 				NumPDBViolations: int64(numPDBViolations),
 			}
-			c := candidate{
+			c := &candidate{
 				victims: &victims,
 				name:    nodeInfoCopy.Node().Name,
 			}
-			candidates = append(candidates, &c)
-			resultLock.Unlock()
+			mu.Lock()
+			if numPDBViolations == 0 {
+				nonViolatingCandidates = append(nonViolatingCandidates, c)
+			} else {
+				violatingCandidates = append(violatingCandidates, c)
+			}
+			if len(nonViolatingCandidates) > 0 {
+				if len(nonViolatingCandidates)+len(violatingCandidates) > int(numCandidates) {
+					cancel()
+				}
+			}
+			mu.Unlock()
+		}
+	}
+	parallelize.Until(parallelCtx, len(potentialNodes), checkNode)
+
+	var candidates []Candidate
+	for _, c := range append(nonViolatingCandidates, violatingCandidates...) {
```

Comment:
actually, if we have nonViolating Candidates... we won't be using the violating ones, so you can discard the later ones.

Reply:
We cannot assert, in this case, violating ones would be useless, as later we will call extenders, and also this function should serve as a common library for both in-tree and out-of-tree plugins - out-of-tree plugins may rely on the candidates and choose a different preemption strategy.

However, the logic can be simplified a bit:

```
candidates := append(nonViolatingCandidates, violatingCandidates...)
if len(candidates) >= numCandidates {
    return candidates[:numCandidates]
}
return candidates
```

Reply:
re @Huang-Wei's suggestion - this wouldn't let the garbage collector reclaim the backing array. For example, let's say `numCandidates` is 100, but we had to dry run preemption on 500 nodes to finally find a non-violating node. This would mean len(candidates) is 500. Even though the returned slice will have a len/cap of 100, the backing array will never be garbage collected and all 500 elements will always be in memory.

Reply:
why do we need to iterate here at all? just return the merged slice, it doesn't really matter if we are returning more than numCandidates.

Reply:
Also, there is the `sort` after this, isn't there? You could have only one slice and just a counter for the non-violating Nodes. And then you can remove the lock.

Reply:
right, you could initialize the slice with numCandidates capacity, and use atomic increments for the index and the non-violating counter, see example here: https://github.com/kubernetes/kubernetes/blob/f99ed92ed149e8cbb09dfcdcff78e6fb4024d3a0/pkg/scheduler/framework/plugins/interpodaffinity/scoring.go#L194

Reply:
Done.